keps/sig-network/0034-20190125-ingress-api-group.md

```diff
@@ -0,0 +1,102 @@
+
+We have a couple of choices (and non-choices, see appendix) for the current resource:
+
+1.  We can delete the current resource from extensions.v1beta1 in anticipation that an improved API can replace it.
+1.  We can copy the API as-is (or with minor changes) into networking.v1beta1, preserving/converting existing data (following the same approach taken with all other extensions.v1beta1 resources). This will allow us to start the cleanup of the extensions API group. This also prepares the API for GA.
+
+Option 1 does not seem realistic in a short-term time frame (a new API will need to be taken through design, alpha/beta/ga phases). At the same time, there are enough users that the existing API cannot be deleted out right.
+
+In terms of moving the API towards GA, the API itself has been available in beta for so long that it has attained defacto GA status through usage and adoption (both by users and by load balancer / ingress controller providers). Abandoning it without a full replacement is not a viable approach.  It is clearly a useful API and captures a non-trivial set of use cases.  At this point, it seems more prudent to declare the current API as something the community will support as a V1, codifying its status, while working on either a V2 Ingress API or an entirely different API with a superset of features.
+
+### Goals
+
+*   Move Ingress to a permanent API group.
+*   Position the Ingress API in a place towards making progress towards GA.
+
+## Proposal
+
+### 1.14
+
+*   Copy the Ingress API to networking.k8s.io/v1beta1 (preserving existing data and round-tripping with the extensions Ingress API, following the approach taken for all other extensions/v1beta1 resources)
+*   Develop a set of planned changes and GA graduation criteria with sig-network (intent is to target a minimal set of bugfixes and non-breaking changes)
+*   Announce extensions/v1beta1 Ingress as deprecated (and announce plan for GA)
+
+### 1.15
+
+*   Update API server to persist in networking.k8s.io/v1beta1
+*   Update in-tree controllers, examples, and clients to target networking.k8s.io/v1beta1
+*   Update Ingress controllers in the kubernetes org to target networking.k8s.io/v1beta1
+*   Update documentation to reference networking.k8s.io/v1beta1
+*   Work to meet graduation criteria and promote API to networking.k8s.io/v1
+*   Announce networking.k8s.io/v1beta1 Ingress as deprecated
+
+### 1.16
+
+*   Update API server to persist in networking.k8s.io/v1
+*   Update in-tree controllers, examples, and clients to target networking.k8s.io/v1
+*   Update Ingress controllers in the kubernetes org to target networking.k8s.io/v1
+*   Update documentation to reference networking.k8s.io/v1
+*   Evangelize availability of v1 Ingress API to out-of-org Ingress controllers
```

Comment:
How out-of-org Ingress controllers will be notified about this change?

Reply:
We will have to do our best to publicize it.

Reply:
similar efforts in the past for integration points with storage or container runtimes have ranged from opening tracking issues in their repos to opening pull requests making updates